File: src/main/scala/com/algorithmlx/inscribers/api/handler/StackHandler.scala
package com.algorithmlx.inscribers.api.handler

import net.minecraft.inventory.{IInventory, Inventory}
import net.minecraft.item.ItemStack
import net.minecraft.util.NonNullList
import net.minecraftforge.items.ItemStackHandler
import org.apache.commons.lang3.ArrayUtils

import java.util
import scala.jdk.FunctionWrappers.AsJavaBiFunction

// Author: BlakeBr0 (https://github.com/BlakeBr0)
// Translate to Scala: AlgorithmLX
class StackHandler(size: Int, changeX: Runnable) extends ItemStackHandler(size) {
  private var slotSize: util.Map[Int, Int] = new util.HashMap()
  private var validator: AsJavaBiFunction[Int, ItemStack, Boolean] = _
  private var maxStack = 64
  private var outputs: Array[Int] = _

  def this(size: Int) = {
    this(size, null)
  }

  override def insertItem(slot: Int, stack: ItemStack, simulate: Boolean): ItemStack = {
    if (this.outputs != null && ArrayUtils.contains(this.outputs, slot)) return stack

    super.insertItem(slot, stack, simulate)
  }

  override def extractItem(slot: Int, amount: Int, simulate: Boolean): ItemStack = {
    if (this.outputs != null && !ArrayUtils.contains(this.outputs, slot)) return ItemStack.EMPTY
    super.extractItem(slot, amount, simulate)
  }

  override def getSlotLimit(slot: Int): Int = if(this.slotSize.containsKey(slot)) this.slotSize.get(slot) else this.maxStack

  override def isItemValid(slot: Int, stack: ItemStack): Boolean = this.validator == null || this.validator.apply(slot, stack)

  override def onContentsChanged(slot: Int): Unit = if (this.changeX != null) this.changeX.run()

  def insert(slot: Int, stack: ItemStack, simulate: Boolean): ItemStack = super.insertItem(slot, stack, simulate)

  def extract(slot: Int, amount: Int, simulate: Boolean): ItemStack = super.extractItem(slot, amount, simulate)

  def getStackList(): NonNullList[ItemStack] = this.stacks

  def getOutputs(): Array[Int] = this.outputs

  def setDefaultSlotLimit(size: Int): Unit = this.maxStack = size

  def addSlotLimit(slot: Int, size: Int): Unit = this.slotSize.put(slot, size)

  def setValidator(validator: AsJavaBiFunction[Int, ItemStack, Boolean]) = this.validator = validator

  def setOutputSlots(slots: Array[Int]): Unit = {
    this.outputs = slots
  }

  def toContainer(): IInventory = {
    JStackHelper.invHelper(this.stacks.toArray(new Array[ItemStack](0)))
  }

  def copy(): StackHandler = {
    val copied = new StackHandler(this.getSlots, this.changeX)
    copied.setDefaultSlotLimit(this.maxStack)
    copied.setValidator(this.validator)
    copied.setOutputSlots(this.outputs)

    this.slotSize.forEach(copied.addSlotLimit)

    var i = 0
    while (i < this.getSlots) {
      val stack = this.getStackInSlot(i)
      copied.setStackInSlot(i, stack.copy)
      i += 1
    }

    copied
  }
}

File: src/main/java/com/algorithmlx/inscribers/api/handler/StackHandler.kt
package com.algorithmlx.inscribers.api.handler

import net.minecraft.inventory.IInventory
import net.minecraft.inventory.Inventory
import net.minecraft.item.ItemStack
import net.minecraft.util.NonNullList
import net.minecraftforge.items.ItemStackHandler
import org.apache.commons.lang3.ArrayUtils

class StackHandler(size: Int, private val contextChanged: Runnable? = null): ItemStackHandler(size) {
    private var slotSize: MutableMap<Int, Int> = mutableMapOf()
    private var validator: ((Int, ItemStack) -> Boolean)? = null
    private var maxStackSize = 64
    private var outputs: IntArray? = null

    override fun insertItem(slot: Int, stack: ItemStack, simulate: Boolean): ItemStack {
        if (this.outputs != null && ArrayUtils.contains(this.outputs, slot)) return stack
        return super.insertItem(slot, stack, simulate)
    }

    override fun extractItem(slot: Int, amount: Int, simulate: Boolean): ItemStack {
        if (this.outputs != null && !ArrayUtils.contains(this.outputs, slot)) return ItemStack.EMPTY
        return super.extractItem(slot, amount, simulate)
    }

    override fun getSlotLimit(slot: Int): Int = if (this.slotSize.containsKey(slot)) this.slotSize[slot]!! else this.maxStackSize

    override fun isItemValid(slot: Int, stack: ItemStack): Boolean = this.validator == null || this.validator!!.invoke(slot, stack)

    fun insert(slot: Int, stack: ItemStack, simulate: Boolean): ItemStack = super.insertItem(slot, stack, simulate)

    fun extract(slot: Int, amount: Int, simulate: Boolean): ItemStack = super.extractItem(slot, amount, simulate)

    fun getStackList(): NonNullList<ItemStack> = this.stacks

    fun getOutputs(): IntArray = this.outputs!!

    fun setDefaultSlotLimit(size: Int) {
        this.maxStackSize = size
    }

    fun addSlotLimit(slot: Int, size: Int) {
        this.slotSize[slot] = size
    }

    fun setValidator(validator: (Int, ItemStack) -> Boolean) {
        this.validator = validator
    }

    fun setOutputSlots(vararg outputs: Int) {
        this.outputs = outputs
    }

    fun toContainer(): IInventory = Inventory(this.stacks.toTypedArray()[0])

    fun copy(): StackHandler {
        val copied = StackHandler(this.slots, this.contextChanged)
        copied.setDefaultSlotLimit(this.maxStackSize)
        copied.setValidator(this.validator!!)
        copied.setOutputSlots(*this.outputs!!)

        this.slotSize.forEach(copied::addSlotLimit)

        for (i in 0 until this.slots) {
            val stack = this.getStackInSlot(i)
            copied.setStackInSlot(i, stack.copy())
        }

        return copied
    }
}